core/trino-main/src/test/java/io/trino/operator/TestExchangeClient.java

```diff
@@ -294,14 +294,14 @@ public void testAddLocation()
         assertTaskIsNotFinished(buffer, task3);
 
         exchangeClient.close();
+        buffer.whenTaskFinished(task3).get(10, SECONDS);
+        assertThat(buffer.getFinishedTasks()).containsExactlyInAnyOrder(task1, task2, task3);
+        assertThat(buffer.getFailedTasks().asMap()).isEmpty();
```

Comment:
With the current implementation it looks like it isn't necessary

Currently the future is assigned before `clientCallback.clientFinished`

```
               synchronized (HttpPageBufferClient.this) {
                    closed = true;
                    if (future == resultFuture) {
                        future = null;
                    }
                    lastUpdate = DateTime.now();
                }
                requestsCompleted.incrementAndGet();
                clientCallback.clientFinished(HttpPageBufferClient.this);
```

However I'm not sure whether it makes sense to implicitly rely on this fact, as it may change in the future and the test may silently become flaky again.